fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PushPredicateThroughJoin.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.doris.nereids.trees.expressions.Literal;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.expressions.SlotExtractor;
+import org.apache.doris.nereids.trees.plans.GroupPlan;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalBinaryPlan;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Push the predicate in the LogicalFilter or LogicalJoin to the join children.
+ * todo: Now, only support eq on condition, support other on condition later
+ */
+public class PushPredicateThroughJoin extends OneRewriteRuleFactory {
+
+    @Override
+    public Rule<Plan> build() {
+        return logicalFilter(logicalJoin()).then(plan -> {
+
+            Expression filterPredicates = plan.operator.getPredicates();
+            Optional<Expression> onPredicates = plan.child().operator.getCondition();
+
+            List<Slot> leftInput = plan.child().left().getOutput();
+            List<Slot> rightInput = plan.child().right().getOutput();
+
+            List<Expression> joinConditions = Lists.newArrayList();
+            List<Expression> otherConjuncts = Lists.newArrayList();
+
+            Expression onCondition = Literal.TRUE_LITERAL;
+            if (onPredicates.isPresent()) {
+                onCondition = onPredicates.get();
+            }
+            ExpressionUtils.extractConjunct(ExpressionUtils.add(onCondition, filterPredicates)).forEach(predicate -> {
+                if (Objects.nonNull(getJoinCondition(predicate, leftInput, rightInput))) {
+                    joinConditions.add(predicate);
+                } else {
+                    otherConjuncts.add(predicate);
+                }
+            });
+
+            List<Expression> leftPredicates = Lists.newArrayList();
+            List<Expression> rightPredicates = Lists.newArrayList();
+
+            for (Expression conjunct : otherConjuncts) {
+                Set<Slot> slots = SlotExtractor.extractSlot(conjunct);
+
+                if (slots.isEmpty()) {
+                    leftPredicates.add(conjunct);
+                    rightPredicates.add(conjunct);
+                    continue;
+                }
+                if (leftInput.containsAll(slots)) {
+                    leftPredicates.add(conjunct);
+                }
+                if (rightInput.containsAll(slots)) {
+                    rightPredicates.add(conjunct);
+                }
+            }
+            otherConjuncts.removeAll(leftPredicates);
+            otherConjuncts.removeAll(rightPredicates);
+
+            joinConditions.addAll(otherConjuncts);
+
+            return pushDownPredicate(plan.child(), joinConditions, leftPredicates, rightPredicates);
```

Comment:
There are actually three predicate pushdowns related to join
1. The filter conditions are pushed up to join
2. The conditions of filter are pushed directly to join
3. The join condition is pushed to the following

where 1 is the basis of 2,3. Your pr only implements 1 and 2.
According to your implementation, the pushdown of these three predicates should be in one Rule rather than split into three Rules. I'm fine with that. 
But you need to clearly represent these three different stages in the code (such as separate function processing).

Reply:
other cases will be added later